### PR TITLE
fix: Support scraping pods metrics that is still in scheduling status and has no assigned node

### PR DESCRIPTION
**Pull request by mickeyzzc**
**What this PR does / why we need it**:

When daemonset is used to scrape the metrics of pods, it can only scrape the metrics of pods assigned to nodes, but in scheduling, for example, pods in pending state cannot be scraped. 
This results in pending pods with no indicators to observe and alert. 
This pr adds a configuration parameter that, when enabled, captures pending pods from apiservice.

### CODE REVIEW

File: examples/daemonsetsharding/deployment-no-node-pods.yaml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        app.kubernetes.io/component: exporter
+        app.kubernetes.io/name: kube-state-metrics
+        app.kubernetes.io/version: 2.10.0
+    spec:
+      automountServiceAccountToken: true
+      containers:
+      - args:
+        - --resources=pods
+        - --node=""
+        - --enable-no-node-scrape
```
Comment (CatherineF-dev) [inline examples/daemonsetsharding/deployment-no-node-pods.yaml:27]: I think we can remove `- --enable-no-node-scrape`. If node == "", then it will collect pod metrics whose don't have  assigned nodes.

**Comment by CatherineF-dev** [comment]
Cool! Thanks for it!
We also found Daemonset sharding loses pod metrics who doesn't have assigned nodes. Just found it can be fixed by this PR.

File: examples/daemonsetsharding/deployment-no-node-pods.yaml (hunk 1 above)
Comment (CatherineF-dev) [reply examples/daemonsetsharding/deployment-no-node-pods.yaml:27]: QQ: could you provide a config example on `only namespace is selected regardless of whether it is assigned to node.`?

Is it --node="" && --namespace="kube-system"

Comment (mickeyzzc) [reply examples/daemonsetsharding/deployment-no-node-pods.yaml:27]: Yes, let me run a test to confirm.

Comment (mickeyzzc) [reply examples/daemonsetsharding/deployment-no-node-pods.yaml:27]: I just ran a test with this PR version.
My cluster currently has the following pods pending
image
When I use the following configuration，it can scrape to pending pods under namespace media
```
          args:
            - '--resources=pods'
            - '--node='
            - '--enable-no-node-scrape'
            - '--namespaces=media'
```
When I use the following configuration，it can scrape to all pods under namespace media
```
          args:
            - '--resources=pods'
            - '--node='
            - '--namespaces=media'
```

Comment (CatherineF-dev) [reply examples/daemonsetsharding/deployment-no-node-pods.yaml:27]: We need to change this:

```
func (n *NodeType) GetNodeFieldSelector() string {
	if string(*n) != "" {
		return fields.OneTermEqualSelector("spec.nodeName", string(*n)).String()
	}
	return EmptyFieldSelector()
}
```
to 
```
func (n *NodeType) GetNodeFieldSelector() string {
	return fields.OneTermEqualSelector("spec.nodeName", string(*n)).String()
}
```

https://github.com/kubernetes/kube-state-metrics/blob/5c71459a614eaf8370bed2f406899f2269edff66/pkg/options/types.go#L111C20-L116

Comment (CatherineF-dev) [reply examples/daemonsetsharding/deployment-no-node-pods.yaml:27]: Oh, I realized the issue.

If we change the above, it will break for `--namespaces=media`. Because we assume collecting all nodes when node = "".

Comment (mickeyzzc) [reply examples/daemonsetsharding/deployment-no-node-pods.yaml:27]: Yes, it is. I think the current PR version is flexible. So this configuration combination was added.

Comment (CatherineF-dev) [reply examples/daemonsetsharding/deployment-no-node-pods.yaml:27]: One solution is to 1. change node type from string to list 2. add an assert that #nodes <= 1:

```
o.cmd.Flags().StringVar((*string)(&o.Node), "node"
```
to 
```
o.cmd.Flags().Var(&o.Namespaces, "namespaces", # similar to namespaces.
```
So that we can differentiate between nil and "". 

https://github.com/kubernetes/kube-state-metrics/blob/5c71459a614eaf8370bed2f406899f2269edff66/pkg/options/options.go#L139C2-L139C2

Comment (CatherineF-dev) [reply examples/daemonsetsharding/deployment-no-node-pods.yaml:27]: QQ: what happened during upgrade? I thought it only had one small breaking change:

Before: --node= works
Now: --node= works

One breaking change: 
Before: --node="" means collecting all nodes.
Now: --node="" means collecting empty node.

But I think most people will not set --node if they don't use this field. We can update release note about it. Since it's an experimental feature, it could have this fix.

I think this way is unified. I didn't realize collecting empty node metrics, and we should fix it.

Comment (mickeyzzc) [reply examples/daemonsetsharding/deployment-no-node-pods.yaml:27]: I'm testing the code, just a little confused to make sure. I will submit the revised version later.

File: examples/daemonsetsharding/deployment-no-node-pods.yaml [hunk 2]
```diff
@@ -0,0 +1,58 @@
+apiVersion: apps/v1
```
Comment (mrueg) [inline examples/daemonsetsharding/deployment-no-node-pods.yaml:1]: This file should be generated from jsonnet See: https://github.com/kubernetes/kube-state-metrics/blob/main/jsonnet/kube-state-metrics/kube-state-metrics.libsonnet

Comment (mickeyzzc) [reply examples/daemonsetsharding/deployment-no-node-pods.yaml:1]: > This file should be generated from jsonnet See: https://github.com/kubernetes/kube-state-metrics/blob/main/jsonnet/kube-state-metrics/kube-state-metrics.libsonnet

fix

File: examples/daemonsetsharding/deployment-no-node-pods.yaml [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+kind: Deployment
+metadata:
+  labels:
+    app.kubernetes.io/component: exporter
+    app.kubernetes.io/name: kube-state-metrics
+    app.kubernetes.io/version: 2.10.0
+  name: kube-state-metrics
```
Comment (CatherineF-dev) [inline examples/daemonsetsharding/deployment-no-node-pods.yaml:8]: nit: kube-state-metrics -> kube-state-metrics-global

We have another deployment with the same name. In the future, we can let ` kube-state-metrics-global` to collect all other metrics which can't be sharded by node.

File: pkg/options/options.go [hunk 4]
```diff
@@ -46,6 +46,7 @@ type Options struct {
 	Namespaces               NamespaceList   `yaml:"namespaces"`
 	NamespacesDenylist       NamespaceList   `yaml:"namespaces_denylist"`
 	Node                     NodeType        `yaml:"node"`
+	NoNodeScrape             bool            `yaml:"no_node_scrape"`
```
Comment (CatherineF-dev) [inline pkg/options/options.go:49]: nit: we don't need this now.

File: pkg/options/types.go [hunk 5]
```diff
@@ -105,14 +106,52 @@ func (r *ResourceSet) Type() string {
 }
 
 // NodeType represents a nodeName to query from.
-type NodeType string
+type NodeType map[string]struct{}
+
+func (n *NodeType) Set(value string) error {
+	s := *n
```
Comment (CatherineF-dev) [inline pkg/options/types.go:112]: Make `s` a new map.

Because if `n` has 10 node names with 2 duplicated.  Length of `cols := make([]string, 0, len(n))` is 10 instead of 9.

File: pkg/options/types.go [hunk 6]
```diff
@@ -105,14 +106,52 @@ func (r *ResourceSet) Type() string {
...
-	if string(*n) != "" {
-		return fields.OneTermEqualSelector("spec.nodeName", string(*n)).String()
+	if nil == n || len(*n) == 0 {
+		klog.InfoS("Using node type is nil")
+		return EmptyFieldSelector()
 	}
-	return EmptyFieldSelector()
+	pattern := "[^a-zA-Z0-9_,-]+"
+	re := regexp.MustCompile(pattern)
+	result := re.ReplaceAllString(n.String(), "")
```
Comment (CatherineF-dev) [inline pkg/options/types.go:147]: QQ: why do we need `re.ReplaceAllString(n.String(), "")`

Comment (CatherineF-dev) [reply pkg/options/types.go:147]: nit: remove `re.ReplaceAllString(n.String(), "")`. If node name isn't correct, keep it as before. Then users can know node name isn't correct.

File: pkg/options/types_test.go [hunk 7]
```diff
@@ -162,13 +162,25 @@ func TestNodeFieldSelector(t *testing.T) {
 		Wanted string
 	}{
 		{
```
Comment (CatherineF-dev) [inline pkg/options/types_test.go:164]: Add a test that it can differentiate between nil and "".

**Comment by CatherineF-dev** [comment]
Cool! Seems we can make it without an extra flag.

File: pkg/options/types_test.go (hunk 7 above)
Comment (mickeyzzc) [reply pkg/options/types_test.go:164]: > ```
> `spec.nodeName=""`  # ""
> `spec.nodeName=` # nil
> ```
> 
> This might work, since we can't use " spec.nodeName="" "

I thought so too, so I added regulars to prevent this, as well as modifications to the configuration.

Comment (mickeyzzc) [reply pkg/options/types_test.go:164]: > ```
> `spec.nodeName=""`  # ""
> `spec.nodeName=` # nil
> ```
> 
> This might work, since we can't use " spec.nodeName="" "

I thought so too, so I added regulars to prevent this, as well as modifications to the configuration.

File: pkg/options/types.go (hunk 6 above)
Comment (mickeyzzc) [reply pkg/options/types.go:147]: > QQ: why we need `re.ReplaceAllString(n.String(), "")`

There are two considerations for regularization here: 
1. Handle the difference between "" and nil 
2. Correction of node name

File: pkg/options/types_test.go (hunk 7 above)
Comment (mickeyzzc) [reply pkg/options/types_test.go:164]: > Add a test that it can differentiate between nil and "".

I fixed it locally. Submit it later along with other questions
image

File: pkg/options/types.go (hunk 5 above)
Comment (mickeyzzc) [reply pkg/options/types.go:112]: > Make `s` a new map.
> 
> Because if `n` has 10 node names with 2 duplicated. Length of `cols := make([]string, 0, len(n))` is 10 instead of 9.

Is it possible to configure the function NewOptions？
```
// NewOptions returns a new instance of `Options`.
func NewOptions() *Options {
	return &Options{
		Resources:            ResourceSet{},
		MetricAllowlist:      MetricSet{},
		MetricDenylist:       MetricSet{},
		MetricOptInList:      MetricSet{},
		Node:                 NodeType{},
		AnnotationsAllowList: LabelsAllowList{},
		LabelsAllowList:      LabelsAllowList{},
	}
}
```

File: examples/daemonsetsharding/deployment-no-node-pods.yaml [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    app.kubernetes.io/component: exporter
+    app.kubernetes.io/name: kube-state-metrics-global
+    app.kubernetes.io/version: 2.10.0
+  name: kube-state-metrics-global
+  namespace: kube-system
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      app.kubernetes.io/name: kube-state-metrics
```
Comment (CatherineF-dev) [inline examples/daemonsetsharding/deployment-no-node-pods.yaml:14]: nit: not sure whether we need to rename it to kube-state-metrics-global.

It has two kube-state-metrics (one Daemonset and one Deployment). If both of them are using the same selector, some wrong pods might be selected. But not sure about it.

File: pkg/options/types.go (hunk 6 above)
Comment (CatherineF-dev) [reply pkg/options/types.go:147]: QQ: could you give two examples for these two cases?

We shouldn't change node name, which users are not aware of.

File: examples/daemonsetsharding/deployment-no-node-pods.yaml (hunk 8 above)
Comment (mickeyzzc) [reply examples/daemonsetsharding/deployment-no-node-pods.yaml:14]: I spent some time testing a few more scenarios to give a multi-point combination of cases out.

Comment (mickeyzzc) [reply examples/daemonsetsharding/deployment-no-node-pods.yaml:14]: Pods will be clearer, I made some combinations, and then came up with some names, it seems that pods will be better, we can make supplementary explanations in the documentation

Comment (mickeyzzc) [reply examples/daemonsetsharding/deployment-no-node-pods.yaml:14]: two kube-state-metrics (one Daemonset and one Deployment).
I tested it directly on my side repeatedly and found no duplication.

**Comment by mickeyzzc** [comment]
> /triage accepted

@dashpole Do I need your approval to merge again?

**Comment by dashpole** [comment]
@mickeyzzc I just did triage, to ensure the PR has an appropriate assignee.  You will need approval from the reviewers and assignees on the PR to merge.

File: pkg/options/types.go (hunk 6 above)
Comment (CatherineF-dev) [reply pkg/options/types.go:150]: Ping~

Keep node name as same from input.

**Comment by mickeyzzc** [comment]
> @mickeyzzc I just did triage, to ensure the PR has an appropriate assignee. You will need approval from the reviewers and assignees on the PR to merge.

@mrueg Can you help with the merger?

**Comment by mrueg** [comment]
To my understanding, when someone runs the daemonset for pod metrics, you suggest to have an additional deployment with --node="" to track unassigned pods?

Isn't it simpler to include pod metrics with --node="" in the regular [ksm deployment](https://github.com/kubernetes/kube-state-metrics/blob/main/examples/daemonsetsharding/deployment.yaml) that is already there additionally to the DS or does it make sense to have an additional deployment? (I'm not sure how users use this in terms of their metric storage system. @CatherineF-dev might be able to provide insights)

And before we can merge, can you update the README so users know how to apply this? 
https://github.com/kubernetes/kube-state-metrics#daemonset-sharding-for-pod-metrics

**Comment by CatherineF-dev** [comment]
> Isn't it simpler to include pod metrics with --node="" in the regular [ksm deployment](https://github.com/kubernetes/kube-state-metrics/blob/main/examples/daemonsetsharding/deployment.yaml) that is already there additionally to the DS or does it make sense to have an additional deployment?

Yes, this PR want to fix ` --node=""` to not select any nodes. Without this CL, --node="" collects all nodes.

**Comment by CatherineF-dev** [comment]
I left one comment here https://github.com/kubernetes/kube-state-metrics/pull/2217/files#r1399988722

**Comment by mickeyzzc** [comment]
> And before we can merge, can you update the README so users know how to apply this?
> https://github.com/kubernetes/kube-state-metrics#daemonset-sharding-for-pod-metrics

@mrueg completed

File: pkg/options/types.go (hunk 6 above)
Comment (mrueg) [reply pkg/options/types.go:150]: @mickeyzzc can you address this comment?

File: README.md [hunk 9]
```diff
@@ -276,6 +276,21 @@ spec:
               fieldPath: spec.nodeName
 ```
 
+To track metrics for unassigned pods, you need to add an additional deployment and set `--node=""`, as shown in the following example:
+```
```
Comment (mrueg) [inline README.md:280]: @mickeyzzc can you add an empty line before the codeblock to fix the lint? Thanks!

File: README.md [hunk 10]
```diff
@@ -276,6 +276,21 @@ spec:
               fieldPath: spec.nodeName
 ```
 
+To track metrics for unassigned pods, you need to add an additional deployment and set `--node=""`, as shown in the following example:
```
Comment (mrueg) [inline README.md:279]:
```suggestion
To track metrics for unassigned pods, you need to add an additional deployment and set `--node=""`, as shown in the following example:

```

**Comment by mrueg** [comment]
/lgtm 

Thank you for your contribution!
